### PR TITLE
add service spec to support override service Spec

### PR DESCRIPTION
**Pull request by TinySong**
fix https://github.com/VictoriaMetrics/operator/issues/163

This feature fix will involve the following components. current just finish vmalert，and vmagent。
add service spec in echo CRD to support override the service spec by user
- [X] vmalert
- [X] vmagent
- [X] vminsert
- [X] vmselect
- [X] vmalertmanager
- [X] vmstorage (no need)
- [X] apidoc
 
there may be some questions:
1. Is this implementation feasible ? 
2. in cluster mode, why vminsert component is deployment resource, and vmselect is statefulset ?
3. if vmalertmanager and vmstorage want to  supports custom service types or other field, need create new  no headless service.

### CODE REVIEW

**Comment by f41gh7** [comment]
My notes:
 1) need to handle case when user changes service type from clusterIP for NodePort for example. In this case old service must be deleted, than new one created.

2) I suggest  keep the current service as is and create separate service with user defined options - service name, labels and annotations. It allows to control loadbalancer behavior only for this service for instance. 

 About questions:
1. implementation looks good, i can suggest to make generic function, that will set default options for service if its not defined for current service, something like: `func setDefaultService(current, default *v1.ServiceSpec) `
2. vminsert is stateless, vmselect may have pvc for cache.
3. separate service for every CRD object should solve this issue.

**Comment by TinySong** [comment]
all component code done.

**Comment by schmurfy** [comment]
While waiting for something like this to be available how can I set the service type for vmsingle to LoadBalancer with fixed address ?
I tried to create another service of that type pointing to the same selectors but it does not seem to find the target, I am not sure if k8s allows having multiple service pointing to the same pods on the same port.

File: api/v1beta1/vmagent_types.go [hunk 1]
```diff
@@ -237,6 +237,9 @@ type VMAgentSpec struct {
 	// ExtraEnvs that will be added to VMAgent pod
 	// +optional
 	ExtraEnvs []v1.EnvVar `json:"extraEnvs,omitempty"`
+	// ServiceSpec that will be added to vmagent service spec
+	// +optional
+	ServiceSpec v1.ServiceSpec `json:"serviceSpec,omitempty"`
```
Comment (f41gh7) [inline api/v1beta1/vmagent_types.go]: Can you add custom type for serviceSpec:

```go
 type ServiceSpec struct {
 Name string
 Spec: v1.ServiceSpec
  Annotations: map[string]string
  Labels: map[string]string
```

 and use it as 
``` go
ServiceSpec *ServiceSpec
```
 for all data structures.

**Comment by f41gh7** [comment]
> While waiting for something like this to be available how can I set the service type for vmsingle to LoadBalancer with fixed address ?
> I tried to create another service of that type pointing to the same selectors but it does not seem to find the target, I am not sure if k8s allows having multiple service pointing to the same pods on the same port.

 It should be ok to create additional services, check labelSelector values, probably you have issue with it.

**Comment by f41gh7** [comment]
@TinySong  sorry for merge conflict, can you resolve it.

File: api/v1beta1/vmagent_types.go (hunk 1 above)
Comment (bobsongplus) [reply api/v1beta1/vmagent_types.go]: The struct looks more reasonable, I will fix it, thanks.

**Comment by TinySong** [comment]
@f41gh7  confliction fixed

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
Thanks for contribution!

**Pull request merged**